### PR TITLE
Config option to disable setting S_ISGID bit when creating installation directory

### PR DESCRIPTION
**Pull request by iarspider**
Addresses spack/spack#14425

### CODE REVIEW

**Comment by iarspider** [comment]
I don't know if there is any other file you need to edit when adding a config option: some JSON schema maybe?

**Comment by adamjstewart** [comment]
There is a JSON schema somewhere, I can probably dig it up if you can't find it. Look for previous PRs that added new config options.

But is there any way we can try-except this, or check if we're on AFS instead of relying on manual user intervention?

**Comment by iarspider** [comment]
There doesn't seem to be an easy reliable way of detecting AFS (or, in general, telling what filesystem is for a given path). As for try-except: it's a valid idea, but will make the code less readable because of nested try-except:
```
try:
    chmod(perms_with_setgid)
except PermissionError:
    try:
        chmod(permis_without_setgid)
    except PermissionError as e:
        raise e
```

An alternative solution could be creating a "probe" directory at the start of `spack install`, trying to set `S_ISGID` flag, and remembering the result - but that is a very extensive change IMO.

**Comment by hartzell** [comment]
Why is Spack setting the `setgid` bit?  On some filesystems it's used to control "BSD" vs "ATT" group semantics:

> When the bit is set for a directory, the set of files in that directory will have the same group as the group of the parent directory, and not that of the user who created those files. This is used for file sharing since they can be now modified by all the users who are part of the group of the parent directory.
>
> from [here on geeksforgeeks](https://www.geeksforgeeks.org/setuid-setgid-and-sticky-bits-in-linux-file-permissions/)

but those permissions are inherited from the parent dir, so if someone *wants* those semantics then they should be able to get them by setting the SETGID bit on the parent directory.

Is there something else going on?

**Comment by iarspider** [comment]
@hartzell @becker33 any progress?

**Comment by hartzell** [comment]
@iarspider -- I'm not a doer on this, just was trying to clear up my confusion about how it was being used.

**Comment by iarspider** [comment]
Ping @becker33 @scheibelp !

File: lib/spack/spack/package_prefs.py [hunk 1]
```diff
@@ -230,7 +230,7 @@ def get_package_dir_permissions(spec):
     attribute sticky for the directory. Package-specific settings take
     precedent over settings for ``all``"""
     perms = get_package_permissions(spec)
-    if perms & stat.S_IRWXG:
+    if perms & stat.S_IRWXG and spack.config.get('config:allow_sgid', True):
```
Comment (becker33) [inline lib/spack/spack/package_prefs.py:233]: We prefer to have default in *both* the code and the `$spack/etc/spack/defaults` config scope. Could you add

`allow_sgid: true`

to `$spack/etc/spack/defaults/config.yaml` along with a comment explaining that this is for filesystems that don't allow sgid manipulation by unprivileged users?

Comment (iarspider) [reply lib/spack/spack/package_prefs.py:233]: @becker33 done

**Comment by becker33** [comment]
Bouncing close/open to restart tests

**Pull request merged**